### PR TITLE
feat: add (File).ReaddirPattern (Share).ReadDirPattern methods

### PR DESCRIPTION
**Pull request by oliverpool**
Hi,

thank you for this library!

Since SMB natively supports reading only a specific pattern while listing the content of a directory, I thought it would be nice to add two methods to the API to support this: `(File).ReaddirPattern` and `(Share).ReadDirPattern` (maybe you have a suggestion for a better name?).

Since those are just additions, this is not a breaking change.

I have refactored the code, so that the methods `Readdir` and `ReadDir` simply call the newly created methods with the `*` pattern.

I would appreciate any comment or suggestions regarding this PR :smile:

### CODE REVIEW

**Comment by oliverpool** [comment]
Apparently when there is no match and the pattern is not `*`, the server will return `STATUS_NO_SUCH_FILE`.

Should this case be silently discarded or should I simply add a comment mentioning this particularity?

**Comment by hirochachacha** [comment]
Hi, thank you for your PR.
I considered this proposal, but I won't accept it as is.
I want to mimic standard library's behavior as precisely as possible in this library.
So, if we want to implement glob feature, I think we should mimic `filepath.Glob` behavior with its name.
Unfortunately, [filepath.Glob](https://pkg.go.dev/path/filepath#Match) and [windows's wildcard pattern](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-cifs/dc92d939-ec45-40c8-96e5-4c4091e4ab43) are not compatible.
Instead of using native pattern matching, we need to copy some code of `filepath` package with some modification.
What do you think?

**Comment by oliverpool** [comment]
> I want to mimic standard library's behavior as precisely as possible in this library.

Fair enough, but as you pointed out, the behaviors of `filepath.Glob` and `smb.pattern` are quite different:
- smb only supports `*` (0 or more char) and `?` (1 char - or 0 at the end of the pattern)
- smb doesn't support any kind of `character-range` (`[a-z]`)

So what you probably mean would be to accept a `filepath.Glob` pattern, translate it somehow to a smb pattern, make the request, filter the reply using `filepath.Glob` to sort out the smb-matched files which do not glob-match.


Example folder
```
abc.ext
ab1.ext
ab9.ext
test
tes
```

**Example 1**
User-request : `ab[0-9].ext`
Smb conversion: `ab?.ext`
Smb matches: `abc.ext`, `ab1.ext`, `ab9.ext`,
glob-filter: `ab1.ext`, `ab9.ext`

**Example 2**
User-request : `tes?`
Smb conversion: `tes?`
Smb matches: `test`, `tes`
glob-filter: `test`


I think this would be quite complex to implement correctly.

---

In my usecase, I have to find some files matching a specific pattern in a big folder (~5000 files) but only a handful do match (less than 100). So having access to this low-level feature would be very nice.

---

I would suggest find a better name than `ReaddirPattern` (`ReaddirWithWildcards` ?) to show that this is very low-level and let the users do the Glob translation themselves (if the folder is not big, just calling `ReadDir` and then `filepath.Glob` should be enough.

I should also add a link to the [relevant smb documentation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-cifs/dc92d939-ec45-40c8-96e5-4c4091e4ab43).

**Comment by hirochachacha** [comment]
Apologies for the delayed response.
I've already finished to make implementation that you suggested one week ago.
but I'm too lazy to test the code. Now I've added test cases.
Can you confirm it solve the problem?

**Comment by oliverpool** [comment]
I just saw your branch, thank you.

I have found an edge case, for which I created a PR: #54 

And I think you currently don't handle the `STATUS_NO_SUCH_FILE` case (which I got from a server, when no match was found for a non `*` pattern).

On the overall, it seems to add a lot of subtle complexity, but if you are fine with maintaining it, it should be fine (for my usecase, I would be happy with the simple exposition of the `pattern` argument and the handling of the `STATUS_NO_SUCH_FILE` error).

**Comment by oliverpool** [comment]
I amended my pull-request to try to get the best of both worlds:
- better document the low-level `ReaddirPattern` methods 
- create a `glob` subpackage, which calls those functions to mimic the `filepath.Glob` behavior (and which contains all the complex subtleties)

This way the base package is kept quite simple and maintainable, while the subpackage contains the complex (and error-prone) part.

What do you think of this compromise?

I did a quick-and-dirty copy of your code in the `glob` subpackage, which must be cleaned-up (and probably renamed) if you choose to go this way.

**Comment by hirochachacha** [comment]
I don't want to expose low-level features in this library.
If I accept this, I should have to accept other low-level features as well for consistency.
I understand that you don't want to use a complex code because you know you can solve the problem simply.
However, mixing low-level and high-level features in one package will probably increase maintenance costs more.
Ideally, we should provide low-level library and rewrite high-level library on top of that, like https://github.com/jborean93/smbprotocol does, so that a power user like you would satisfy as well as a novice user.
But it takes too much time to do that, so it's not our option now.
In conclusion, I still feel like providing `Glob` function in the same library is the best option.

**Comment by oliverpool** [comment]
Fair enough!

I totally understand your arguments and I agree with you, that 

> Ideally, we should provide low-level library and rewrite high-level library on top of that

But also know the effort which would be required by such a big change.

I will try your `feat/glob` branch, if it works fine in my case (can you open a PR with this branch, so that we can continue the discussion there?)

**Comment by oliverpool** [comment]
I just realized that [`Share.Glob`](https://pkg.go.dev/github.com/hirochachacha/go-smb2#Share.Glob) is available since v1.1.0. Thanks !
